// @alexaegis/autotool-plugin-eslint@0.8.4/index.js
import { not, or, equal, contains } from "@alexaegis/predicate";
import "autotool-plugin";
import { join } from "node:path";
const name = "@alexaegis/autotool-plugin-eslint";
const description = "Setup eslint";
const version = "0.8.4";
const license = "MIT";
const archetype = {
  platform: "node",
  framework: "autotool",
  language: "ts",
  kind: "lib"
};
const keywords = [
  "autotool-plugin",
  "eslint",
  "managed-by-autotool"
];
const type = "module";
const publishConfig = {
  access: "public"
};
const scripts = {
  build: "turbo run build-lib_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-eslint",
  "build-lib_": "vite build",
  "lint:depcheck": "turbo run lint:depcheck_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-eslint",
  "lint:depcheck_": "depcheck",
  "lint:es": "turbo run lint:es_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-eslint",
  "lint:es_": "eslint --max-warnings=0 --fix --no-error-on-unmatched-pattern .",
  "lint:format": "turbo run lint:format_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-eslint",
  "lint:format_": "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --check .",
  "lint:md": "turbo run lint:md_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-eslint",
  "lint:md_": "remark --frail --no-stdout --silently-ignore .",
  "lint:tsc": "turbo run lint:tsc_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-eslint",
  "lint:tsc_": "tsc --noEmit",
  publint: "BUILD_REASON='publish' turbo run publint_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-eslint",
  publint_: "publint dist",
  all: "BUILD_REASON='publish' turbo run all_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-eslint",
  format: "turbo run format_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-eslint",
  format_: "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --write .",
  test: "turbo run test_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-eslint",
  test_: "vitest --passWithNoTests --coverage --run",
  "test:watch": "vitest --passWithNoTests --coverage"
};
const exports = {
  ".": {
    types: "./src/index.ts",
    "import": "./dist/index.js",
    require: "./dist/index.cjs",
    "default": "./dist/index.js"
  },
  "./eslintignore": "./static/eslintignore.txt",
  "./package-eslintrc.cjs": "./static/package-eslintrc.cjs.txt",
  "./readme": "./readme.md",
  "./workspace-eslintrc": "./static/workspace-eslintrc.cjs"
};
const dependencies = {
  "@alexaegis/logging": "^0.7.2",
  "@alexaegis/predicate": "^0.7.2",
  "autotool-plugin": "^0.3.0"
};
const devDependencies = {
  "@alexaegis/eslint-config-core": "^0.8.4",
  "@alexaegis/eslint-config-vitest": "^0.8.4",
  "@alexaegis/ts": "^0.8.4",
  "@alexaegis/vite": "^0.8.4",
  "@alexaegis/vitest": "^0.8.4",
  "@types/eslint": "^8.44.7",
  "@types/node": "^20.9.0",
  eslint: "^8.53.0",
  publint: "^0.2.5",
  typescript: "^5.2.2",
  vite: "^4.5.0",
  vitest: "^0.34.6"
};
const packageJson = {
  name,
  description,
  version,
  license,
  "private": false,
  archetype,
  keywords,
  type,
  publishConfig,
  scripts,
  exports,
  dependencies,
  devDependencies
};
const joinAdditionalExtends = (...pluginNames) => {
  return pluginNames.length > 0 ? `, '${pluginNames.join("', '")}'` : "";
};
const plugin = (_options) => {
  return {
    name: packageJson.name,
    elements: [
      {
        description: "workspace eslint config",
        executor: "fileCopy",
        packageKind: "root",
        formatWithPrettier: true,
        targetFile: ".eslintrc.cjs",
        sourcePluginPackageName: packageJson.name,
        sourceFile: join("static", "workspace-eslintrc.cjs")
      },
      {
        description: "workspace eslintignore",
        executor: "fileCopy",
        packageKind: "root",
        targetFile: ".eslintignore",
        sourcePluginPackageName: packageJson.name,
        sourceFile: join("static", "eslintignore.txt")
      },
      {
        description: "package eslintignore",
        executor: "fileCopy",
        packageKind: "regular",
        targetFile: ".eslintignore",
        sourcePluginPackageName: packageJson.name,
        sourceFile: join("static", "eslintignore.txt")
      },
      {
        description: "package eslint config (non svelte)",
        executor: "fileCopy",
        packageKind: "regular",
        formatWithPrettier: true,
        targetFile: ".eslintrc.cjs",
        packageJsonFilter: {
          name: not(or(equal("@alexaegis/vite"), equal("@alexaegis/vitest"))),
          // Skip these libraries to avoid circular dependencies
          archetype: {
            framework: not(contains("svelte"))
          }
        },
        sourcePluginPackageName: packageJson.name,
        templateVariables: {
          additionalExtends: joinAdditionalExtends("@alexaegis/eslint-config-vitest")
        },
        sourceFile: join("static", "package-eslintrc.cjs.txt")
      },
      {
        description: "package eslint config (svelte)",
        executor: "fileCopy",
        packageKind: "regular",
        formatWithPrettier: true,
        targetFile: ".eslintrc.cjs",
        packageJsonFilter: {
          name: not(or(equal("@alexaegis/vite"), equal("@alexaegis/vitest"))),
          // Skip these libraries to avoid circular dependencies
          archetype: {
            framework: "svelte"
          }
        },
        sourcePluginPackageName: packageJson.name,
        templateVariables: {
          additionalExtends: joinAdditionalExtends(
            "@alexaegis/eslint-config-svelte",
            "@alexaegis/eslint-config-vitest"
          )
        },
        sourceFile: join("static", "package-eslintrc.cjs.txt")
      },
      {
        description: "package eslint dependencies (vitest)",
        executor: "packageJson",
        packageKind: "regular",
        packageJsonFilter: {
          name: /^(?!@alexaegis\/(eslint-config-)?vite(st)?).*$/
        },
        data: {
          devDependencies: {
            "@alexaegis/eslint-config-vitest": packageJson.devDependencies["@alexaegis/eslint-config-vitest"]
          }
        }
      },
      {
        description: "package eslint dependencies (svelte)",
        executor: "packageJson",
        packageKind: "regular",
        packageJsonFilter: {
          archetype: {
            framework: "svelte"
          }
        },
        data: {
          devDependencies: {
            "@alexaegis/eslint-config-svelte": `^${packageJson.version}`
            // Versioned together, it's fine
          }
        }
      },
      {
        description: "package eslint config (For @alexaegis/vitest only)",
        executor: "fileCopy",
        packageKind: "regular",
        formatWithPrettier: true,
        targetFile: ".eslintrc.cjs",
        packageJsonFilter: {
          name: /^@alexaegis\/vite(st)?$/
        },
        templateVariables: {
          additionalExtends: joinAdditionalExtends()
          // No extra config here
        },
        sourcePluginPackageName: packageJson.name,
        sourceFile: join("static", "package-eslintrc.cjs.txt")
      },
      {
        description: "workspace eslint dependencies",
        executor: "packageJson",
        packageKind: "root",
        data: {
          devDependencies: {
            "@alexaegis/eslint-config-core": packageJson.devDependencies["@alexaegis/eslint-config-core"],
            eslint: packageJson.devDependencies.eslint,
            "@types/eslint": packageJson.devDependencies["@types/eslint"]
          }
        }
      },
      {
        description: "workspace eslint scripts",
        executor: "packageJson",
        packageKind: "root",
        data: {
          scripts: {
            "lint:es": "turbo run lint:es_ --concurrency 16 --cache-dir .cache/turbo"
          }
        }
      },
      {
        description: "package eslint scripts",
        executor: "packageJson",
        packageKind: "regular",
        data: {
          scripts: {
            "lint:es": "turbo run lint:es_ --concurrency 16 --cache-dir .cache/turbo --filter ${packageName}",
            "lint:es_": "eslint --max-warnings=0 --fix --no-error-on-unmatched-pattern ."
          }
        }
      }
    ]
  };
};
export {
  plugin as default,
  joinAdditionalExtends,
  plugin
};
