// @alexaegis/coverage-tools@0.7.3/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const common = require("@alexaegis/common");
const workspaceTools = require("@alexaegis/workspace-tools");
const globby = require("globby");
const promises = require("node:fs/promises");
const normalizeCollectLcovReportPathsOptions = (options) => {
  const { onlyWorkspaceRoot, skipWorkspaceRoot, ...rest } = workspaceTools.normalizeCollectWorkspacePackagesOptions(options);
  return rest;
};
const LCOV_INFO_FILE_NAME = "lcov.info";
const collectLcovReportPaths = async (rawOptions) => {
  const options = normalizeCollectLcovReportPathsOptions(rawOptions);
  const workspaceRoot = workspaceTools.getWorkspaceRoot(options);
  if (common.isNullish(workspaceRoot)) {
    return [];
  }
  const workspacePackages = await workspaceTools.collectWorkspacePackages({
    ...options,
    skipWorkspaceRoot: true
  });
  const lcovPathResults = await Promise.all(
    workspacePackages.map(
      (workspacePackage) => globby.globby([`${workspacePackage.packagePath}/**/${LCOV_INFO_FILE_NAME}`], {
        absolute: true,
        onlyFiles: true,
        cwd: workspaceRoot,
        ignore: [`**/${workspaceTools.NODE_MODULES_DIRECTORY_NAME}`]
      })
    )
  );
  return lcovPathResults.flat();
};
const mergeLcovReportsInWorkspace = async (rawOptions) => {
  const options = workspaceTools.normalizeCollectWorkspacePackagesOptions(rawOptions);
  const lcovPaths = await collectLcovReportPaths(rawOptions);
  options.logger.info(
    `found the following lcov files in the workpace:
	- ${lcovPaths.join("\n	- ")}`
  );
  const allLcovReports = await common.asyncFilterMap(
    lcovPaths,
    (path) => promises.readFile(path, {
      encoding: "utf8"
    }).catch(() => void 0)
  );
  return allLcovReports.join("\n");
};
exports.LCOV_INFO_FILE_NAME = LCOV_INFO_FILE_NAME;
exports.collectLcovReportPaths = collectLcovReportPaths;
exports.mergeLcovReportsInWorkspace = mergeLcovReportsInWorkspace;
