// @alexaegis/autotool-plugin-remark@0.8.4/index.js
import "autotool-plugin";
import { join } from "node:path";
const name = "@alexaegis/autotool-plugin-remark";
const description = "My markdown linting config";
const version = "0.8.4";
const license = "MIT";
const archetype = {
  platform: "node",
  framework: "autotool",
  language: "ts",
  kind: "lib"
};
const keywords = [
  "autotool-plugin",
  "managed-by-autotool",
  "markdown"
];
const type = "module";
const publishConfig = {
  access: "public"
};
const scripts = {
  build: "turbo run build-lib_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-remark",
  "build-lib_": "vite build",
  "lint:depcheck": "turbo run lint:depcheck_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-remark",
  "lint:depcheck_": "depcheck",
  "lint:es": "turbo run lint:es_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-remark",
  "lint:es_": "eslint --max-warnings=0 --fix --no-error-on-unmatched-pattern .",
  "lint:format": "turbo run lint:format_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-remark",
  "lint:format_": "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --check .",
  "lint:md": "turbo run lint:md_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-remark",
  "lint:md_": "remark --frail --no-stdout --silently-ignore .",
  "lint:tsc": "turbo run lint:tsc_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-remark",
  "lint:tsc_": "tsc --noEmit",
  publint: "BUILD_REASON='publish' turbo run publint_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-remark",
  publint_: "publint dist",
  all: "BUILD_REASON='publish' turbo run all_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-remark",
  format: "turbo run format_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-remark",
  format_: "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --write .",
  test: "turbo run test_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-remark",
  test_: "vitest --passWithNoTests --coverage --run",
  "test:watch": "vitest --passWithNoTests --coverage"
};
const exports = {
  ".": {
    types: "./src/index.ts",
    "import": "./dist/index.js",
    require: "./dist/index.cjs",
    "default": "./dist/index.js"
  },
  "./dot-remarkignore": "./static/dot-remarkignore",
  "./dot-remarkrc": "./static/dot-remarkrc.js",
  "./readme": "./readme.md",
  "./remark-preset": {
    types: "./src/remark-preset.ts",
    "import": "./dist/remark-preset.js",
    require: "./dist/remark-preset.cjs",
    "default": "./dist/remark-preset.js"
  }
};
const dependencies = {
  "@alexaegis/logging": "^0.7.2",
  "@alexaegis/remark-preset": "^0.8.4",
  "autotool-plugin": "^0.3.0"
};
const devDependencies = {
  "@alexaegis/eslint-config-vitest": "^0.8.4",
  "@alexaegis/ts": "^0.8.4",
  "@alexaegis/vite": "^0.8.4",
  "@alexaegis/vitest": "^0.8.4",
  "@types/node": "^20.9.0",
  publint: "^0.2.5",
  remark: "^15.0.1",
  "remark-cli": "^12.0.0",
  vite: "^4.5.0",
  vitest: "^0.34.6"
};
const packageJson = {
  name,
  description,
  version,
  license,
  "private": false,
  archetype,
  keywords,
  type,
  publishConfig,
  scripts,
  exports,
  dependencies,
  devDependencies
};
const plugin = (_options) => {
  return {
    name: packageJson.name,
    elements: [
      {
        description: "workspace remark scripts and devDependencies",
        executor: "packageJson",
        packageKind: "root",
        data: {
          scripts: {
            "lint:md": "turbo run lint:md_ --concurrency 16 --cache-dir .cache/turbo",
            "lint:md_": "remark --frail --no-stdout --silently-ignore --ignore-pattern packages --ignore-pattern libs --ignore-pattern apps ."
            // ? I would've like to use a .config/workspace-only.remarkignore file but remark didn't use it when passed with --ignore-path, maybe a bug?
          },
          devDependencies: {
            "remark-cli": packageJson.devDependencies["remark-cli"],
            remark: packageJson.devDependencies.remark
            // A local installation is required by the remark langauge server
          }
        }
      },
      {
        description: "package remark scripts",
        executor: "packageJson",
        packageKind: "regular",
        data: {
          scripts: {
            "lint:md": "turbo run lint:md_ --concurrency 16 --cache-dir .cache/turbo --filter ${packageName}",
            "lint:md_": "remark --frail --no-stdout --silently-ignore ."
          }
        }
      },
      {
        description: "workspace remarkignore file",
        executor: "fileCopy",
        packageKind: "root",
        sourceFile: join("static", "dot-remarkignore"),
        targetFile: ".remarkignore",
        sourcePluginPackageName: packageJson.name
      },
      {
        description: "workspace remarkrc config file",
        executor: "fileCopy",
        packageKind: "root",
        formatWithPrettier: true,
        sourceFile: join("static", "dot-remarkrc.js"),
        targetFile: ".remarkrc.js",
        sourcePluginPackageName: packageJson.name
      }
    ]
  };
};
export {
  plugin as default,
  plugin
};
