// @alexaegis/svelte-config@0.8.4/internal/configurations.cjs
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const vite = require('@sveltejs/kit/vite');
const autoProcess_js = require('svelte-preprocess/dist/autoProcess.js');
const pickAdapter_function = require('./pick-adapter.function.cjs');
const toBaseHref_function = require('./to-base-href.function.cjs');

/**
 * Adds the regular sveltePreprocessors
 *
 * @type {import('./types.js').Config}
 */
const libConfiguration = {
	preprocess: autoProcess_js.sveltePreprocess(),
};

/**
 * Enables the vite inspector on `shift`
 *
 * @type {import('./types.js').Config}
 */
const spaConfiguration = {
	...libConfiguration,
	vitePlugin: {
		inspector: {
			holdMode: true,
			toggleKeyCombo: 'shift',
		},
	},
};

/**
 * Enables the vite inspector on `shift`, overrides the regular svelte preprocessor with kit's
 * And sets the base href to the value of `process.env['BASE_HREF']`
 * And the platform based on `process.env['PLATFORM']`
 * Valid values are 'vercel', 'github-pages' and 'auto' which is the default
 *
 * @type {import('./types.js').Config}
 */
const kitConfiguration = {
	...spaConfiguration,
	preprocess: vite.vitePreprocess(),
	kit: {
		paths: {
			base: toBaseHref_function.toBaseHref(process.env['BASE_HREF']),
		},
		adapter: pickAdapter_function.pickAdapter(pickAdapter_function.normalizePlatform(process.env['PLATFORM'])),
	},
};

exports.kitConfiguration = kitConfiguration;
exports.libConfiguration = libConfiguration;
exports.spaConfiguration = spaConfiguration;
