// @alexaegis/vite@0.8.4/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const vite = require("vite");
const vitePluginPakk = require("vite-plugin-pakk");
const logging = require("@alexaegis/logging");
const isAbsolute = (path) => {
  return path.startsWith("/");
};
const toBaseHref = (path = "") => {
  const trimmed = path.replace(/\/$/, "");
  if (trimmed) {
    return isAbsolute(trimmed) ? trimmed : `/${trimmed}`;
  } else {
    return "";
  }
};
const DEFAULT_OUT_DIR = "dist";
const DEFAULT_ENTRY = ["src/index.ts"];
const DEFAULT_EXPORT_FORMATS = ["es", "cjs"];
const DEFAULT_BUILD_TARGET = "es2022";
const DEFAULT_VITE_CONFIG = {
  build: {
    target: DEFAULT_BUILD_TARGET,
    outDir: DEFAULT_OUT_DIR
  }
};
const DEFAULT_VITE_APP_CONFIG = vite.mergeConfig(DEFAULT_VITE_CONFIG, {
  base: toBaseHref(process.env["BASE_HREF"])
});
const DEFAULT_VITE_LIB_CONFIG = vite.mergeConfig(DEFAULT_VITE_CONFIG, {
  build: {
    minify: false,
    sourcemap: true,
    rollupOptions: {
      external: vitePluginPakk.createLazyAutoExternalsFunction(),
      // I'm always using this, but autolib also adds it with the other defaults if they are not defined
      treeshake: true
    },
    lib: {
      entry: DEFAULT_ENTRY,
      formats: DEFAULT_EXPORT_FORMATS
    }
  }
});
const DEFAULT_VITE_JS_LIB_CONFIG = vite.mergeConfig(DEFAULT_VITE_LIB_CONFIG, {
  build: {
    rollupOptions: {
      output: {
        preserveModules: true
        // Otherwise type paths would be mangled
      }
    }
  },
  esbuild: false
  // esbuild always removes comments, and JSDoc wouldn't work
});
const defineAppConfig = (config) => vite.mergeConfig(DEFAULT_VITE_CONFIG, config);
const defineLibConfig = (config) => vite.mergeConfig(DEFAULT_VITE_LIB_CONFIG, config);
const disabledPlugin = (disabledPluginName) => {
  const name = `disabled:${disabledPluginName}`;
  const logger = logging.createLogger({
    name: `vite:${name}`
  });
  return {
    name,
    buildStart: () => {
      logger.info(`Skipping plugin ${disabledPluginName}`);
    }
  };
};
const conditionalPlugin = (plugin, condition) => {
  if (condition()) {
    return plugin;
  } else {
    let name = "unknown";
    if (typeof plugin === "object" && plugin.name) {
      name = plugin.name;
    }
    return disabledPlugin(name);
  }
};
const isTargetEnvNotLocal = () => process.env["TARGET_ENV"]?.toLowerCase() !== "local";
const libraryViteConfig = defineLibConfig({
  plugins: [conditionalPlugin(vitePluginPakk.pakk(), isTargetEnvNotLocal)]
});
Object.defineProperty(exports, "pakk", {
  enumerable: true,
  get: () => vitePluginPakk.pakk
});
exports.DEFAULT_BUILD_TARGET = DEFAULT_BUILD_TARGET;
exports.DEFAULT_ENTRY = DEFAULT_ENTRY;
exports.DEFAULT_EXPORT_FORMATS = DEFAULT_EXPORT_FORMATS;
exports.DEFAULT_OUT_DIR = DEFAULT_OUT_DIR;
exports.DEFAULT_VITE_APP_CONFIG = DEFAULT_VITE_APP_CONFIG;
exports.DEFAULT_VITE_CONFIG = DEFAULT_VITE_CONFIG;
exports.DEFAULT_VITE_JS_LIB_CONFIG = DEFAULT_VITE_JS_LIB_CONFIG;
exports.DEFAULT_VITE_LIB_CONFIG = DEFAULT_VITE_LIB_CONFIG;
exports.conditionalPlugin = conditionalPlugin;
exports.defineAppConfig = defineAppConfig;
exports.defineLibConfig = defineLibConfig;
exports.disabledPlugin = disabledPlugin;
exports.isAbsolute = isAbsolute;
exports.isTargetEnvNotLocal = isTargetEnvNotLocal;
exports.libraryViteConfig = libraryViteConfig;
exports.toBaseHref = toBaseHref;
