// @pakk/cli@0.7.0/package-7ee32a93.js
const name = "@pakk/cli";
const description = "The standalone cli of pakk, that can manage your package.json for library development.";
const version = "0.7.0";
const license = "MIT";
const archetype = {
  platform: "node",
  language: "ts",
  kind: "lib"
};
const keywords = [
  "javascript",
  "js",
  "managed-by-autotool",
  "ts",
  "typescript"
];
const homepage = "https://www.github.com/AlexAegis/pakk#readme";
const type = "module";
const config = {
  "engine-strict": false
};
const publishConfig = {
  access: "public"
};
const scripts = {
  build: "turbo run build-lib_ --concurrency 16 --cache-dir .cache/turbo --filter @pakk/cli",
  "build-lib_": "vite build",
  "lint:depcheck": "turbo run lint:depcheck_ --concurrency 16 --cache-dir .cache/turbo --filter @pakk/cli",
  "lint:depcheck_": "depcheck",
  "lint:es": "turbo run lint:es_ --concurrency 16 --cache-dir .cache/turbo --filter @pakk/cli",
  "lint:es_": "eslint --max-warnings=0 --fix --no-error-on-unmatched-pattern .",
  "lint:format": "turbo run lint:format_ --concurrency 16 --cache-dir .cache/turbo --filter @pakk/cli",
  "lint:format_": "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --check .",
  "lint:md": "turbo run lint:md_ --concurrency 16 --cache-dir .cache/turbo --filter @pakk/cli",
  "lint:md_": "remark --frail --no-stdout --silently-ignore .",
  "lint:tsc": "turbo run lint:tsc_ --concurrency 16 --cache-dir .cache/turbo --filter @pakk/cli",
  "lint:tsc_": "tsc --noEmit",
  publint: "BUILD_REASON='publish' turbo run publint_ --concurrency 16 --cache-dir .cache/turbo --filter @pakk/cli",
  publint_: "publint dist",
  all: "BUILD_REASON='publish' turbo run all_ --concurrency 16 --cache-dir .cache/turbo --filter @pakk/cli",
  format: "turbo run format_ --concurrency 16 --cache-dir .cache/turbo --filter @pakk/cli",
  format_: "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --write .",
  test: "turbo run test_ --concurrency 16 --cache-dir .cache/turbo --filter @pakk/cli",
  test_: "vitest --passWithNoTests --coverage --run",
  "test:watch": "vitest --passWithNoTests --coverage"
};
const exports = {
  ".": {
    types: "./src/index.ts",
    "import": "./src/index.js",
    require: "./src/index.cjs",
    "default": "./src/index.js"
  },
  "./readme": "./readme.md"
};
const bin = {
  pakk: "./shims/pakk.js",
  "sort-json": "./shims/sort-json.js"
};
const dependencies = {
  "@alexaegis/cli-tools": "^0.7.0",
  "@alexaegis/common": "^0.7.0",
  "@alexaegis/fs": "^0.7.0",
  "@alexaegis/logging": "^0.7.0",
  "@alexaegis/workspace-tools": "^0.7.0",
  "@pakk/core": "^0.7.0",
  "@types/node": "^20.5.7",
  prettier: "^3.0.3",
  "ts-node": "^10.9.1",
  typescript: "^5.2.2",
  vite: "^4.4.9",
  yargs: "^17.7.2"
};
const devDependencies = {
  "@alexaegis/eslint-config-vitest": "^0.8.0",
  "@alexaegis/ts": "^0.8.0",
  "@alexaegis/vite": "^0.8.0",
  "@alexaegis/vitest": "^0.8.0",
  "@types/node": "^20.5.7",
  "@types/yargs": "^17.0.24",
  publint: "^0.2.2",
  vite: "^4.4.9",
  "vite-plugin-pakk": "^0.7.0",
  vitest: "^0.34.3"
};
const packageJson = {
  name,
  description,
  version,
  license,
  "private": false,
  archetype,
  keywords,
  homepage,
  type,
  config,
  publishConfig,
  scripts,
  exports,
  bin,
  dependencies,
  devDependencies
};
export {
  packageJson as p
};
