// @alexaegis/autotool-plugin-typedoc@0.8.4/index.cjs
"use strict";
Object.defineProperties(exports, { __esModule: { value: true }, [Symbol.toStringTag]: { value: "Module" } });
require("autotool-plugin");
const node_path = require("node:path");
const name = "@alexaegis/autotool-plugin-typedoc";
const description = "My typedoc config";
const version = "0.8.4";
const license = "MIT";
const archetype = {
  platform: "node",
  framework: "autotool",
  language: "ts",
  kind: "lib"
};
const keywords = [
  "autotool-plugin",
  "managed-by-autotool",
  "typedoc"
];
const type = "module";
const publishConfig = {
  access: "public"
};
const scripts = {
  build: "turbo run build-lib_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-typedoc",
  "build-lib_": "vite build",
  "lint:depcheck": "turbo run lint:depcheck_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-typedoc",
  "lint:depcheck_": "depcheck",
  "lint:es": "turbo run lint:es_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-typedoc",
  "lint:es_": "eslint --max-warnings=0 --fix --no-error-on-unmatched-pattern .",
  "lint:format": "turbo run lint:format_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-typedoc",
  "lint:format_": "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --check .",
  "lint:md": "turbo run lint:md_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-typedoc",
  "lint:md_": "remark --frail --no-stdout --silently-ignore .",
  "lint:tsc": "turbo run lint:tsc_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-typedoc",
  "lint:tsc_": "tsc --noEmit",
  publint: "BUILD_REASON='publish' turbo run publint_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-typedoc",
  publint_: "publint dist",
  all: "BUILD_REASON='publish' turbo run all_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-typedoc",
  format: "turbo run format_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-typedoc",
  format_: "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --write .",
  test: "turbo run test_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-typedoc",
  test_: "vitest --passWithNoTests --coverage --run",
  "test:watch": "vitest --passWithNoTests --coverage"
};
const exports$1 = {
  ".": {
    types: "./src/index.ts",
    "import": "./dist/index.js",
    require: "./dist/index.cjs",
    "default": "./dist/index.js"
  },
  "./package-typedoc": "./static/package-typedoc.json",
  "./readme": "./readme.md",
  "./typedoc.base": "./static/typedoc.base.json",
  "./workspace-typedoc": "./static/workspace-typedoc.json"
};
const dependencies = {
  "@alexaegis/logging": "^0.7.2",
  "autotool-plugin": "^0.3.0",
  typedoc: "^0.25.3"
};
const devDependencies = {
  "@alexaegis/eslint-config-vitest": "^0.8.4",
  "@alexaegis/ts": "^0.8.4",
  "@alexaegis/vite": "^0.8.4",
  "@alexaegis/vitest": "^0.8.4",
  "@types/node": "^20.9.0",
  publint: "^0.2.5",
  typescript: "^5.2.2",
  vite: "^4.5.0",
  vitest: "^0.34.6"
};
const packageJson = {
  name,
  description,
  version,
  license,
  "private": false,
  archetype,
  keywords,
  type,
  publishConfig,
  scripts,
  exports: exports$1,
  dependencies,
  devDependencies
};
const plugin = (_options) => {
  const packageJsonFilter = {
    archetype: {
      language: /^(ts|js)$/
      // Typedoc can process JSDoc too
    }
  };
  return {
    name: packageJson.name,
    elements: [
      {
        description: "workspace typedoc scripts and devDependencies",
        executor: "packageJson",
        packageKind: "root",
        data: {
          scripts: {
            typedoc: "turbo run typedoc_ --concurrency 16 --cache-dir .cache/turbo",
            typedoc_: "typedoc"
          },
          devDependencies: {
            typedoc: packageJson.dependencies.typedoc
          }
        }
      },
      {
        description: "package typedoc config file",
        executor: "fileCopy",
        packageJsonFilter,
        packageKind: "regular",
        formatWithPrettier: true,
        sourceFile: node_path.join("static", "package-typedoc.json"),
        targetFile: "typedoc.json",
        sourcePluginPackageName: packageJson.name
      },
      {
        description: "workspace typedoc config file",
        executor: "fileCopy",
        packageKind: "root",
        formatWithPrettier: true,
        sourceFile: node_path.join("static", "workspace-typedoc.json"),
        targetFile: "typedoc.json",
        sourcePluginPackageName: packageJson.name
      },
      {
        description: "workspace base typedoc config file",
        executor: "fileCopy",
        packageKind: "root",
        sourceFile: node_path.join("static", "typedoc.base.json"),
        targetFile: node_path.join(".config", "typedoc.base.json"),
        sourcePluginPackageName: packageJson.name
      }
    ]
  };
};
exports.default = plugin;
exports.plugin = plugin;
