// @alexaegis/autotool-plugin-vitest@0.8.4/index.cjs
"use strict";
Object.defineProperties(exports, { __esModule: { value: true }, [Symbol.toStringTag]: { value: "Module" } });
const predicate = require("@alexaegis/predicate");
require("autotool-plugin");
const node_path = require("node:path");
const name = "@alexaegis/autotool-plugin-vitest";
const description = "Setup vitest";
const version = "0.8.4";
const license = "MIT";
const archetype = {
  platform: "node",
  framework: "autotool",
  language: "ts",
  kind: "lib"
};
const keywords = [
  "autotool-plugin",
  "managed-by-autotool",
  "vitest"
];
const type = "module";
const publishConfig = {
  access: "public"
};
const scripts = {
  build: "turbo run build-lib_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-vitest",
  "build-lib_": "vite build",
  "lint:depcheck": "turbo run lint:depcheck_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-vitest",
  "lint:depcheck_": "depcheck",
  "lint:es": "turbo run lint:es_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-vitest",
  "lint:es_": "eslint --max-warnings=0 --fix --no-error-on-unmatched-pattern .",
  "lint:format": "turbo run lint:format_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-vitest",
  "lint:format_": "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --check .",
  "lint:md": "turbo run lint:md_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-vitest",
  "lint:md_": "remark --frail --no-stdout --silently-ignore .",
  "lint:tsc": "turbo run lint:tsc_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-vitest",
  "lint:tsc_": "tsc --noEmit",
  publint: "BUILD_REASON='publish' turbo run publint_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-vitest",
  publint_: "publint dist",
  all: "BUILD_REASON='publish' turbo run all_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-vitest",
  format: "turbo run format_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-vitest",
  format_: "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --write .",
  test: "turbo run test_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-vitest",
  test_: "vitest --passWithNoTests --coverage --run",
  "test:watch": "vitest --passWithNoTests --coverage"
};
const exports$1 = {
  ".": {
    types: "./src/index.ts",
    "import": "./dist/index.js",
    require: "./dist/index.cjs",
    "default": "./dist/index.js"
  },
  "./package-node-vitest.config": "./static/package-node-vitest.config.ts",
  "./package-svelte-vitest.config": "./static/package-svelte-vitest.config.ts",
  "./package-web-vitest.config": "./static/package-web-vitest.config.ts",
  "./readme": "./readme.md"
};
const dependencies = {
  "@alexaegis/coverage-tools": "^0.7.2",
  "@alexaegis/logging": "^0.7.2",
  "@alexaegis/predicate": "^0.7.2",
  "autotool-plugin": "^0.3.0"
};
const devDependencies = {
  "@alexaegis/eslint-config-vitest": "^0.8.4",
  "@alexaegis/ts": "^0.8.4",
  "@alexaegis/vite": "^0.8.4",
  "@alexaegis/vitest": "^0.8.4",
  "@lcov-viewer/cli": "^1.3.0",
  "@types/node": "^20.9.0",
  "@vitest/coverage-v8": "^0.34.6",
  publint: "^0.2.5",
  typescript: "^5.2.2",
  vite: "^4.5.0",
  vitest: "^0.34.6"
};
const packageJson = {
  name,
  description,
  version,
  license,
  "private": false,
  archetype,
  keywords,
  type,
  publishConfig,
  scripts,
  exports: exports$1,
  dependencies,
  devDependencies
};
const appOrLib = predicate.or(predicate.equal("app"), predicate.equal("lib"));
const plugin = (_options) => {
  return {
    name: packageJson.name,
    elements: [
      {
        description: "workspace test scripts and devDependencies",
        executor: "packageJson",
        packageKind: "root",
        data: {
          scripts: {
            test: "turbo run test_ --concurrency 16 --cache-dir .cache/turbo && merge-workspace-lcov-reports && lcov-viewer lcov -o ./coverage ./coverage/lcov.info"
          },
          devDependencies: {
            "@vitest/coverage-v8": packageJson.devDependencies["@vitest/coverage-v8"],
            "@lcov-viewer/cli": packageJson.devDependencies["@lcov-viewer/cli"],
            "@alexaegis/coverage-tools": packageJson.dependencies["@alexaegis/coverage-tools"]
          }
        }
      },
      {
        description: "package test scripts",
        executor: "packageJson",
        packageKind: "regular",
        packageJsonFilter: {
          archetype: {
            kind: appOrLib
          }
        },
        data: {
          scripts: {
            test: "turbo run test_ --concurrency 16 --cache-dir .cache/turbo --filter ${packageName}",
            test_: "vitest --passWithNoTests --coverage --run",
            "test:watch": "vitest --passWithNoTests --coverage"
          }
        }
      },
      {
        description: "package test devDependencies",
        executor: "packageJson",
        packageKind: "regular",
        packageJsonFilter: {
          name: predicate.not(predicate.or(predicate.equal("@alexaegis/vite"), predicate.equal("@alexaegis/vitest"))),
          // Don't add it for itself and 'vite' it would cause a circle
          archetype: {
            kind: appOrLib
          }
        },
        data: {
          devDependencies: {
            "@alexaegis/vitest": packageJson.devDependencies["@alexaegis/vitest"],
            vitest: packageJson.devDependencies.vitest
          }
        }
      },
      {
        description: "package vitest config file for web packages",
        executor: "fileCopy",
        packageJsonFilter: {
          archetype: {
            kind: appOrLib,
            platform: "web",
            framework: predicate.not(predicate.contains("svelte"))
          }
        },
        packageKind: "regular",
        formatWithPrettier: true,
        sourceFile: node_path.join("static", "package-web-vitest.config.ts"),
        targetFile: "vitest.config.ts",
        sourcePluginPackageName: packageJson.name
      },
      {
        description: "package vitest config file for node packages",
        executor: "fileCopy",
        packageJsonFilter: {
          archetype: {
            kind: appOrLib,
            platform: "node",
            framework: predicate.not(predicate.contains("svelte"))
          }
        },
        packageKind: "regular",
        formatWithPrettier: true,
        sourceFile: node_path.join("static", "package-node-vitest.config.ts"),
        targetFile: "vitest.config.ts",
        sourcePluginPackageName: packageJson.name
      },
      {
        description: "package vitest config file for svelte packages",
        executor: "fileCopy",
        packageJsonFilter: {
          archetype: {
            kind: appOrLib,
            framework: "svelte"
          }
        },
        packageKind: "regular",
        formatWithPrettier: true,
        sourceFile: node_path.join("static", "package-svelte-vitest.config.ts"),
        targetFile: "vitest.config.ts",
        sourcePluginPackageName: packageJson.name
      }
    ]
  };
};
exports.default = plugin;
exports.plugin = plugin;
