// autotool@0.3.0/index.js
import { r, q, c, a, l, d, e, s, f, t, g, i, u, h, b, k, j, m, w, v, n, o, p } from "./autotool.function-40ea481f.js";
import { relative } from "node:path";
import "@alexaegis/common";
import "@alexaegis/workspace-tools";
import "autotool-plugin";
import "node:child_process";
import "node:fs";
import "@alexaegis/match";
import "globby";
import "@alexaegis/fs";
import "microdiff";
import "autotool-plugin-default";
const relativePathBetweenPackages = (from, to) => {
  return relative(from.packagePath, to.packagePath) || "./";
};
export {
  r as autotool,
  q as autotoolPluginFilterPredicate,
  c as checkIfTheresAnElementWithoutValidExecutor,
  a as consolidateElementsAndFilterOutNonExecutables,
  l as createIsPackageElementTargeted,
  d as discoverPackageManager,
  e as elementAndPluginFilter,
  s as executeElementsOnPackage,
  f as filterElementsForPackage,
  t as findInstalledPlugins,
  g as groupAndConsolidateElementsByTargetFile,
  i as installCommands,
  u as isAutotoolPluginObject,
  h as isElementTargeted,
  b as isElementUntargeted,
  k as isPackageElementTargeted,
  j as isPackageElementUntargeted,
  m as isRootWorkspacePackage,
  w as loadInstalledPlugins,
  v as loadPlugin,
  n as mapRecord,
  o as normalizeElementTargets,
  p as partition,
  relativePathBetweenPackages
};
