// autotool@0.3.0/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const autotool_function = require("./autotool.function-22a64c29.cjs");
const node_path = require("node:path");
require("@alexaegis/common");
require("@alexaegis/workspace-tools");
require("autotool-plugin");
require("node:child_process");
require("node:fs");
require("@alexaegis/match");
require("globby");
require("@alexaegis/fs");
require("microdiff");
require("autotool-plugin-default");
const relativePathBetweenPackages = (from, to) => {
  return node_path.relative(from.packagePath, to.packagePath) || "./";
};
exports.autotool = autotool_function.autotool;
exports.autotoolPluginFilterPredicate = autotool_function.autotoolPluginFilterPredicate;
exports.checkIfTheresAnElementWithoutValidExecutor = autotool_function.checkIfTheresAnElementWithoutValidExecutor;
exports.consolidateElementsAndFilterOutNonExecutables = autotool_function.consolidateElementsAndFilterOutNonExecutables;
exports.createIsPackageElementTargeted = autotool_function.createIsPackageElementTargeted;
exports.discoverPackageManager = autotool_function.discoverPackageManager;
exports.elementAndPluginFilter = autotool_function.elementAndPluginFilter;
exports.executeElementsOnPackage = autotool_function.executeElementsOnPackage;
exports.filterElementsForPackage = autotool_function.filterElementsForPackage;
exports.findInstalledPlugins = autotool_function.findInstalledPlugins;
exports.groupAndConsolidateElementsByTargetFile = autotool_function.groupAndConsolidateElementsByTargetFile;
exports.installCommands = autotool_function.installCommands;
exports.isAutotoolPluginObject = autotool_function.isAutotoolPluginObject;
exports.isElementTargeted = autotool_function.isElementTargeted;
exports.isElementUntargeted = autotool_function.isElementUntargeted;
exports.isPackageElementTargeted = autotool_function.isPackageElementTargeted;
exports.isPackageElementUntargeted = autotool_function.isPackageElementUntargeted;
exports.isRootWorkspacePackage = autotool_function.isRootWorkspacePackage;
exports.loadInstalledPlugins = autotool_function.loadInstalledPlugins;
exports.loadPlugin = autotool_function.loadPlugin;
exports.mapRecord = autotool_function.mapRecord;
exports.normalizeElementTargets = autotool_function.normalizeElementTargets;
exports.partition = autotool_function.partition;
exports.relativePathBetweenPackages = relativePathBetweenPackages;
