// @alexaegis/autotool-plugin-prettier@0.8.4/index.js
import "autotool-plugin";
import { join } from "node:path";
const name = "@alexaegis/autotool-plugin-prettier";
const description = "My prettier config distributor";
const version = "0.8.4";
const license = "MIT";
const archetype = {
  platform: "node",
  framework: "autotool",
  language: "ts",
  kind: "lib"
};
const keywords = [
  "autotool-plugin",
  "managed-by-autotool",
  "prettier"
];
const type = "module";
const publishConfig = {
  access: "public"
};
const scripts = {
  build: "turbo run build-lib_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-prettier",
  "build-lib_": "vite build",
  "lint:depcheck": "turbo run lint:depcheck_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-prettier",
  "lint:depcheck_": "depcheck",
  "lint:es": "turbo run lint:es_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-prettier",
  "lint:es_": "eslint --max-warnings=0 --fix --no-error-on-unmatched-pattern .",
  "lint:format": "turbo run lint:format_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-prettier",
  "lint:format_": "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --check .",
  "lint:md": "turbo run lint:md_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-prettier",
  "lint:md_": "remark --frail --no-stdout --silently-ignore .",
  "lint:tsc": "turbo run lint:tsc_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-prettier",
  "lint:tsc_": "tsc --noEmit",
  publint: "BUILD_REASON='publish' turbo run publint_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-prettier",
  publint_: "publint dist",
  all: "BUILD_REASON='publish' turbo run all_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-prettier",
  format: "turbo run format_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-prettier",
  format_: "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --write .",
  test: "turbo run test_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-prettier",
  test_: "vitest --passWithNoTests --coverage --run",
  "test:watch": "vitest --passWithNoTests --coverage"
};
const exports = {
  ".": {
    types: "./src/index.ts",
    "import": "./dist/index.js",
    require: "./dist/index.cjs",
    "default": "./dist/index.js"
  },
  "./prettier-config": {
    types: "./src/prettier-config.ts",
    "import": "./dist/prettier-config.js",
    require: "./dist/prettier-config.cjs",
    "default": "./dist/prettier-config.js"
  },
  "./prettierrc.js": "./static/prettierrc.js.txt",
  "./readme": "./readme.md",
  "./workspace": "./static/workspace.prettierignore",
  "./workspace-only": "./static/workspace-only.prettierignore"
};
const dependencies = {
  "@alexaegis/logging": "^0.7.2",
  "@alexaegis/prettier-config": "^0.8.4",
  "autotool-plugin": "^0.3.0",
  prettier: "^3.1.0",
  "prettier-plugin-svelte": "^3.1.0",
  "prettier-plugin-tailwindcss": "^0.5.7"
};
const devDependencies = {
  "@alexaegis/eslint-config-vitest": "^0.8.4",
  "@alexaegis/ts": "^0.8.4",
  "@alexaegis/vite": "^0.8.4",
  "@alexaegis/vitest": "^0.8.4",
  "@types/node": "^20.9.0",
  publint: "^0.2.5",
  typescript: "^5.2.2",
  vite: "^4.5.0",
  vitest: "^0.34.6"
};
const packageJson = {
  name,
  description,
  version,
  license,
  "private": false,
  archetype,
  keywords,
  type,
  publishConfig,
  scripts,
  exports,
  dependencies,
  devDependencies
};
const plugin = (_options) => {
  return {
    name: packageJson.name,
    elements: [
      {
        // ? This and the like could be split into two elements, one for turbo scripts, filtered to workspaces where turbo is used, but right now it's always used so whatever
        description: "add package formatting scripts",
        executor: "packageJson",
        packageKind: "regular",
        data: {
          scripts: {
            format: "turbo run format_ --concurrency 16 --cache-dir .cache/turbo --filter ${packageName}",
            format_: "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --write .",
            // TODO: Remove plugin arguments from here once https://github.com/prettier/prettier/issues/15079 is resolved
            "lint:format_": "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --check .",
            "lint:format": "turbo run lint:format_ --concurrency 16 --cache-dir .cache/turbo --filter ${packageName}"
          }
        }
      },
      {
        description: "add formatting scripts and prettier as a devDependency to the workspace",
        executor: "packageJson",
        packageKind: "root",
        data: {
          scripts: {
            format: "turbo run format_ --concurrency 16 --cache-dir .cache/turbo",
            format_: "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --ignore-path .config/workspace-only.prettierignore --write .",
            // TODO: Remove plugin arguments from here once https://github.com/prettier/prettier/issues/15079 is resolved
            "lint:format_": "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --ignore-path .config/workspace-only.prettierignore --check .",
            // Only check files not under a package
            "lint:format": "turbo run lint:format_ --concurrency 16 --cache-dir .cache/turbo"
          },
          devDependencies: {
            prettier: packageJson.dependencies.prettier,
            // Since the plugins are set as string in the distributed configuration these have to be installed somewhere where prettier can find them
            "prettier-plugin-svelte": packageJson.dependencies["prettier-plugin-svelte"],
            "prettier-plugin-tailwindcss": packageJson.dependencies["prettier-plugin-tailwindcss"]
          }
        }
      },
      {
        description: "copy workspace prettier ignore file used by scripts",
        executor: "fileCopy",
        packageKind: "root",
        sourcePluginPackageName: packageJson.name,
        sourceFile: join("static", "workspace-only.prettierignore"),
        targetFile: join(".config", "workspace-only.prettierignore")
      },
      {
        description: "copy workspace prettier config",
        executor: "fileCopy",
        packageKind: "root",
        formatWithPrettier: true,
        targetFile: ".prettierrc.js",
        sourcePluginPackageName: packageJson.name,
        sourceFile: join("static", "prettierrc.js.txt")
      },
      {
        description: "remove old cjs prettierrc files",
        executor: "fileRemove",
        targetFile: ".prettierrc.cjs"
      },
      {
        description: "remove package config overrides as they arent supported",
        executor: "fileRemove",
        packageKind: "regular",
        targetFile: ".prettierrc.js"
      },
      {
        description: "copy workspace prettierignore used by the editor",
        executor: "fileCopy",
        packageKind: "root",
        targetFile: ".prettierignore",
        sourcePluginPackageName: packageJson.name,
        sourceFile: join("static", "workspace.prettierignore")
      },
      {
        description: "copy package prettierignore",
        executor: "fileCopy",
        // TODO: change this to 'fileRemove' once this is resolved https://github.com/prettier/prettier/issues/4081
        packageKind: "regular",
        targetFile: ".prettierignore",
        sourcePluginPackageName: packageJson.name,
        sourceFile: join("static", "workspace.prettierignore")
      }
    ]
  };
};
export {
  plugin as default,
  plugin
};
