// @alexaegis/autotool-plugin-husky@0.8.4/index.cjs
"use strict";
Object.defineProperties(exports, { __esModule: { value: true }, [Symbol.toStringTag]: { value: "Module" } });
require("autotool-plugin");
const node_path = require("node:path");
const lintStagedConfig = require("@alexaegis/lint-staged-config");
const name = "@alexaegis/autotool-plugin-husky";
const description = "Husky, lint-staged and commitlint stuff";
const version = "0.8.4";
const license = "MIT";
const archetype = {
  platform: "node",
  framework: "autotool",
  language: "ts",
  kind: "lib"
};
const keywords = [
  "autotool-plugin",
  "husky",
  "managed-by-autotool"
];
const type = "module";
const publishConfig = {
  access: "public"
};
const scripts = {
  build: "turbo run build-lib_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-husky",
  "build-lib_": "vite build",
  "lint:depcheck": "turbo run lint:depcheck_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-husky",
  "lint:depcheck_": "depcheck",
  "lint:es": "turbo run lint:es_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-husky",
  "lint:es_": "eslint --max-warnings=0 --fix --no-error-on-unmatched-pattern .",
  "lint:format": "turbo run lint:format_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-husky",
  "lint:format_": "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --check .",
  "lint:md": "turbo run lint:md_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-husky",
  "lint:md_": "remark --frail --no-stdout --silently-ignore .",
  "lint:tsc": "turbo run lint:tsc_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-husky",
  "lint:tsc_": "tsc --noEmit",
  publint: "BUILD_REASON='publish' turbo run publint_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-husky",
  publint_: "publint dist",
  all: "BUILD_REASON='publish' turbo run all_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-husky",
  format: "turbo run format_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-husky",
  format_: "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --write .",
  test: "turbo run test_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-husky",
  test_: "vitest --passWithNoTests --coverage --run",
  "test:watch": "vitest --passWithNoTests --coverage"
};
const exports$1 = {
  ".": {
    types: "./src/index.ts",
    "import": "./dist/index.js",
    require: "./dist/index.cjs",
    "default": "./dist/index.js"
  },
  "./commit-msg": "./static/.husky/commit-msg",
  "./commitlint.config": "./static/commitlint.config.cjs",
  "./lint-staged.config": "./static/lint-staged.config.js",
  "./pre-commit": "./static/.husky/pre-commit",
  "./readme": "./readme.md"
};
const dependencies = {
  "@alexaegis/lint-staged-config": "^0.8.4",
  "@alexaegis/logging": "^0.7.2",
  "@commitlint/config-conventional": "^18.4.0",
  "autotool-plugin": "^0.3.0",
  husky: "^8.0.3",
  "lint-staged": "^15.1.0"
};
const devDependencies = {
  "@alexaegis/eslint-config-vitest": "^0.8.4",
  "@alexaegis/ts": "^0.8.4",
  "@alexaegis/vite": "^0.8.4",
  "@alexaegis/vitest": "^0.8.4",
  "@types/node": "^20.9.0",
  publint: "^0.2.5",
  typescript: "^5.2.2",
  vite: "^4.5.0",
  vitest: "^0.34.6"
};
const packageJson = {
  name,
  description,
  version,
  license,
  "private": false,
  archetype,
  keywords,
  type,
  publishConfig,
  scripts,
  exports: exports$1,
  dependencies,
  devDependencies
};
const plugin = (_options) => {
  return {
    name: packageJson.name,
    elements: [
      {
        description: "add husky as a workspace devDependenecy",
        executor: "packageJson",
        packageKind: "root",
        data: {
          devDependencies: {
            husky: packageJson.dependencies.husky
          }
        }
      },
      {
        description: "add lint-staged as a workspace devDependenecy",
        executor: "packageJson",
        packageKind: "root",
        data: {
          devDependencies: {
            "lint-staged": packageJson.dependencies["lint-staged"]
          }
        }
      },
      {
        description: "root lint-staged config",
        executor: "fileCopy",
        packageKind: "root",
        formatWithPrettier: true,
        targetFile: "lint-staged.config.js",
        sourcePluginPackageName: packageJson.name,
        sourceFile: node_path.join("static", "lint-staged.config.js")
      },
      {
        description: "remove mjs root lint-staged config in favor of js",
        // TODO: remove once propagated
        executor: "fileRemove",
        packageKind: "root",
        targetFile: "lint-staged.config.mjs"
      },
      {
        description: "husky commit-msg hook",
        executor: "fileCopy",
        packageKind: "root",
        targetFile: node_path.join(".husky", "commit-msg"),
        sourcePluginPackageName: packageJson.name,
        sourceFile: node_path.join("static", ".husky", "commit-msg"),
        markAsExecutable: true
      },
      {
        description: "husky pre-commit hook",
        executor: "fileCopy",
        packageKind: "root",
        targetFile: node_path.join(".husky", "pre-commit"),
        sourcePluginPackageName: packageJson.name,
        sourceFile: node_path.join("static", ".husky", "pre-commit"),
        markAsExecutable: true
      }
    ]
  };
};
exports.default = plugin;
exports.plugin = plugin;
Object.keys(lintStagedConfig).forEach((k) => {
  if (k !== "default" && !Object.prototype.hasOwnProperty.call(exports, k))
    Object.defineProperty(exports, k, {
      enumerable: true,
      get: () => lintStagedConfig[k]
    });
});
