// @alexaegis/autotool-plugin-workspace@0.8.4/index.js
import "autotool-plugin";
import { join } from "node:path";
const name$1 = "@alexaegis/workspace";
const description$1 = "base repo for common stuff in my other js projects";
const version$1 = "0.8.4";
const license$1 = "mit";
const keywords$1 = [
  "eslint",
  "javascript",
  "js",
  "managed-by-autotool",
  "ts",
  "tsconfig",
  "turbo",
  "typescript"
];
const author = {
  email: "alexaegis@gmail.com",
  name: "Alex Aegis",
  url: "https://github.com/AlexAegis"
};
const homepage = "https://github.com/AlexAegis/js-tooling";
const repository = {
  url: "git+https://github.com/AlexAegis/js-tooling",
  type: "git"
};
const bugs = {
  email: "alexaegis@gmail.com",
  url: "https://github.com/AlexAegis/js-tooling/issues"
};
const type$1 = "module";
const config = {
  "engine-strict": true
};
const engines = {
  node: ">=18.10.0",
  pnpm: ">=8.0.0"
};
const packageManager = "pnpm@8.10.4";
const workspaces = [
  "apps/*",
  "fixtures/*",
  "libs/*",
  "packages/*"
];
const scripts$1 = {
  build: "turbo run build-lib_ build-app_ --concurrency 16 --cache-dir .cache/turbo",
  "build-app": "turbo run build-app_ --concurrency 16 --cache-dir .cache/turbo",
  "build-lib": "turbo run build-lib_ --concurrency 16 --cache-dir .cache/turbo",
  lint: "turbo run lint_ --concurrency 16 --cache-dir .cache/turbo",
  "lint:depcheck": "turbo run lint:depcheck_ --concurrency 16 --cache-dir .cache/turbo",
  "lint:depcheck_": "depcheck",
  "lint:es": "turbo run lint:es_ --concurrency 16 --cache-dir .cache/turbo",
  "lint:format": "turbo run lint:format_ --concurrency 16 --cache-dir .cache/turbo",
  "lint:format_": "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --ignore-path .config/workspace-only.prettierignore --check .",
  "lint:md": "turbo run lint:md_ --concurrency 16 --cache-dir .cache/turbo",
  "lint:md_": "remark --frail --no-stdout --silently-ignore --ignore-pattern packages --ignore-pattern libs --ignore-pattern apps .",
  "lint:style": "turbo run lint:style_ --concurrency 16 --cache-dir .cache/turbo",
  "lint:style:css": "turbo run lint:style:css_ --concurrency 16 --cache-dir .cache/turbo",
  "lint:style:html": "turbo run lint:style:html_ --concurrency 16 --cache-dir .cache/turbo",
  "lint:style:scss": "turbo run lint:style:scss_ --concurrency 16 --cache-dir .cache/turbo",
  "lint:svelte": "turbo run lint:svelte_ --concurrency 16 --cache-dir .cache/turbo",
  "lint:tsc": "turbo run lint:tsc_ --concurrency 16 --cache-dir .cache/turbo",
  publint: "BUILD_REASON='publish' turbo run publint_ --concurrency 16 --cache-dir .cache/turbo",
  all: "BUILD_REASON='publish' turbo run all_ --concurrency 16 --cache-dir .cache/turbo",
  dev: "pnpm run --dir packages/turbowatch build && pnpm run --dir apps/svelte-example-app dev",
  format: "turbo run format_ --concurrency 16 --cache-dir .cache/turbo",
  format_: "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --ignore-path .config/workspace-only.prettierignore --write .",
  ncu: "ncu --configFilePath ./.config/ && pnpm up",
  nuke: "nuke",
  postinstall: "husky install && turbo run build-lib_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/eslint-config* --filter @alexaegis/prettier-config* --filter @alexaegis/autotool-plugin-* && autotool",
  release: "commit-and-tag-version --sign --commit-all",
  "release:major": "commit-and-tag-version --sign --commit-all --release-as major",
  "release:minor": "commit-and-tag-version --sign --commit-all --release-as minor",
  "release:patch": "commit-and-tag-version --sign --commit-all --release-as patch",
  start: "pnpm run --dir apps/svelte-example-app start",
  test: "turbo run test_ --concurrency 16 --cache-dir .cache/turbo && merge-workspace-lcov-reports && lcov-viewer lcov -o ./coverage ./coverage/lcov.info",
  typedoc: "turbo run typedoc_ --concurrency 16 --cache-dir .cache/turbo",
  typedoc_: "typedoc"
};
const devDependencies$1 = {
  "@alexaegis/autotool-plugin-commitlint": "workspace:^",
  "@alexaegis/autotool-plugin-depcheck": "workspace:^",
  "@alexaegis/autotool-plugin-editorconfig": "workspace:^",
  "@alexaegis/autotool-plugin-eslint": "workspace:^",
  "@alexaegis/autotool-plugin-git": "workspace:^",
  "@alexaegis/autotool-plugin-github": "workspace:^",
  "@alexaegis/autotool-plugin-husky": "workspace:^",
  "@alexaegis/autotool-plugin-ncu": "workspace:^",
  "@alexaegis/autotool-plugin-prettier": "workspace:^",
  "@alexaegis/autotool-plugin-publint": "workspace:^",
  "@alexaegis/autotool-plugin-remark": "workspace:^",
  "@alexaegis/autotool-plugin-standard-version": "workspace:^",
  "@alexaegis/autotool-plugin-stylelint": "workspace:^",
  "@alexaegis/autotool-plugin-svelte": "workspace:^",
  "@alexaegis/autotool-plugin-ts": "workspace:^",
  "@alexaegis/autotool-plugin-turbo": "workspace:^",
  "@alexaegis/autotool-plugin-typedoc": "workspace:^",
  "@alexaegis/autotool-plugin-vite": "workspace:^",
  "@alexaegis/autotool-plugin-vitest": "workspace:^",
  "@alexaegis/autotool-plugin-vscode": "workspace:^",
  "@alexaegis/autotool-plugin-workspace": "workspace:^",
  "@alexaegis/coverage-tools": "^0.7.2",
  "@alexaegis/eslint-config-core": "workspace:^",
  "@alexaegis/nuke": "workspace:^",
  "@alexaegis/stylelint-config": "workspace:^",
  "@alexaegis/ts": "workspace:^",
  "@lcov-viewer/cli": "^1.3.0",
  "@types/eslint": "^8.44.7",
  "@types/node": "^20.9.0",
  "@types/yargs": "^17.0.31",
  "@vitest/coverage-v8": "^0.34.6",
  autotool: "^0.3.0",
  "commit-and-tag-version": "^12.0.0",
  commitlint: "^18.4.1",
  depcheck: "^1.4.7",
  eslint: "^8.53.0",
  husky: "^8.0.3",
  "lint-staged": "^15.1.0",
  "npm-check-updates": "^16.14.6",
  prettier: "^3.1.0",
  "prettier-plugin-svelte": "^3.1.0",
  "prettier-plugin-tailwindcss": "^0.5.7",
  publint: "^0.2.5",
  remark: "^15.0.1",
  "remark-cli": "^12.0.0",
  stylelint: "^15.11.0",
  "svelte-check": "^3.6.0",
  "ts-node": "^10.9.1",
  turbo: "^1.10.16",
  "turbo-ignore": "^1.10.16",
  typedoc: "^0.25.3",
  typescript: "^5.2.2",
  vitest: "^0.34.6"
};
const rootPackageJson = {
  name: name$1,
  description: description$1,
  version: version$1,
  license: license$1,
  "private": true,
  keywords: keywords$1,
  author,
  homepage,
  repository,
  bugs,
  type: type$1,
  config,
  engines,
  packageManager,
  workspaces,
  scripts: scripts$1,
  devDependencies: devDependencies$1
};
const name = "@alexaegis/autotool-plugin-workspace";
const description = "This package contains generic workspace tools and scripts";
const version = "0.8.4";
const license = "MIT";
const archetype = {
  platform: "node",
  framework: "autotool",
  language: "ts",
  kind: "lib"
};
const keywords = [
  "autotool-plugin",
  "managed-by-autotool",
  "workspace"
];
const type = "module";
const publishConfig = {
  access: "public"
};
const scripts = {
  build: "turbo run build-lib_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-workspace",
  "build-lib_": "vite build",
  "lint:depcheck": "turbo run lint:depcheck_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-workspace",
  "lint:depcheck_": "depcheck",
  "lint:es": "turbo run lint:es_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-workspace",
  "lint:es_": "eslint --max-warnings=0 --fix --no-error-on-unmatched-pattern .",
  "lint:format": "turbo run lint:format_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-workspace",
  "lint:format_": "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --check .",
  "lint:md": "turbo run lint:md_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-workspace",
  "lint:md_": "remark --frail --no-stdout --silently-ignore .",
  "lint:tsc": "turbo run lint:tsc_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-workspace",
  "lint:tsc_": "tsc --noEmit",
  publint: "BUILD_REASON='publish' turbo run publint_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-workspace",
  publint_: "publint dist",
  all: "BUILD_REASON='publish' turbo run all_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-workspace",
  format: "turbo run format_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-workspace",
  format_: "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --write .",
  test: "turbo run test_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-workspace",
  test_: "vitest --passWithNoTests --coverage --run",
  "test:watch": "vitest --passWithNoTests --coverage"
};
const exports = {
  ".": {
    types: "./src/index.ts",
    "import": "./dist/index.js",
    require: "./dist/index.cjs",
    "default": "./dist/index.js"
  },
  "./package.sort": "./static/package.sort.json",
  "./pnpm-workspace": "./static/pnpm-workspace.yaml",
  "./readme": "./readme.md"
};
const dependencies = {
  "@alexaegis/logging": "^0.7.2",
  "autotool-plugin": "^0.3.0"
};
const devDependencies = {
  "@alexaegis/eslint-config-vitest": "^0.8.4",
  "@alexaegis/nuke": "^0.8.4",
  "@alexaegis/ts": "^0.8.4",
  "@alexaegis/vite": "^0.8.4",
  "@alexaegis/vitest": "^0.8.4",
  "@types/node": "^20.9.0",
  publint: "^0.2.5",
  typescript: "^5.2.2",
  vite: "^4.5.0",
  vitest: "^0.34.6"
};
const packageJson = {
  name,
  description,
  version,
  license,
  "private": false,
  archetype,
  keywords,
  type,
  publishConfig,
  scripts,
  exports,
  dependencies,
  devDependencies
};
const plugin = (_options) => {
  return {
    name: packageJson.name,
    elements: [
      {
        description: "pnpm workspace preferences",
        executor: "fileCopy",
        packageKind: "root",
        sourcePluginPackageName: packageJson.name,
        sourceFile: join("static", "pnpm-workspace.yaml"),
        targetFile: "pnpm-workspace.yaml"
      },
      {
        description: "package.json sort preferences",
        executor: "fileCopy",
        packageKind: "root",
        sourcePluginPackageName: packageJson.name,
        sourceFile: join("static", "package.sort.json"),
        targetFile: join(".config", "package.sort.json")
      },
      {
        description: "workspace engine preferences for node",
        executor: "packageJson",
        packageKind: "root",
        data: {
          config: {
            "engine-strict": true
          },
          engines: {
            node: ">=18.10.0",
            npm: void 0,
            pnpm: ">=8.0.0"
          },
          packageManager: rootPackageJson.packageManager,
          workspaces: ["apps/*", "libs/*", "packages/*", "fixtures/*"]
          // Some tools like typedoc will only read this field even through the same information is defined in the pnpm-workspace.yaml
        }
      },
      {
        description: "workspace scripts and all autotool plugins as devDependencies",
        executor: "packageJson",
        packageKind: "root",
        data: {
          scripts: {
            nuke: "nuke"
          },
          devDependencies: {
            "@alexaegis/nuke": packageJson.devDependencies["@alexaegis/nuke"],
            "@alexaegis/autotool-plugin-commitlint": `^${packageJson.version}`,
            // versioned together
            "@alexaegis/autotool-plugin-depcheck": `^${packageJson.version}`,
            // versioned together
            "@alexaegis/autotool-plugin-editorconfig": `^${packageJson.version}`,
            // versioned together
            "@alexaegis/autotool-plugin-eslint": `^${packageJson.version}`,
            // versioned together
            "@alexaegis/autotool-plugin-git": `^${packageJson.version}`,
            // versioned together
            "@alexaegis/autotool-plugin-github": `^${packageJson.version}`,
            // versioned together
            "@alexaegis/autotool-plugin-husky": `^${packageJson.version}`,
            // versioned together
            "@alexaegis/autotool-plugin-ncu": `^${packageJson.version}`,
            // versioned together
            "@alexaegis/autotool-plugin-prettier": `^${packageJson.version}`,
            // versioned together
            "@alexaegis/autotool-plugin-publint": `^${packageJson.version}`,
            // versioned together
            "@alexaegis/autotool-plugin-remark": `^${packageJson.version}`,
            // versioned together
            "@alexaegis/autotool-plugin-standard-version": `^${packageJson.version}`,
            // versioned together
            "@alexaegis/autotool-plugin-stylelint": `^${packageJson.version}`,
            // versioned together
            "@alexaegis/autotool-plugin-svelte": `^${packageJson.version}`,
            // versioned together
            "@alexaegis/autotool-plugin-ts": `^${packageJson.version}`,
            // versioned together
            "@alexaegis/autotool-plugin-turbo": `^${packageJson.version}`,
            // versioned together
            "@alexaegis/autotool-plugin-typedoc": `^${packageJson.version}`,
            // versioned together
            "@alexaegis/autotool-plugin-vite": `^${packageJson.version}`,
            // versioned together
            "@alexaegis/autotool-plugin-vitest": `^${packageJson.version}`,
            // versioned together
            "@alexaegis/autotool-plugin-vscode": `^${packageJson.version}`,
            // versioned together
            "@alexaegis/autotool-plugin-workspace": `^${packageJson.version}`
            // versioned together
          }
        }
      }
    ]
  };
};
export {
  plugin as default,
  plugin
};
