// @alexaegis/autotool-plugin-vscode@0.8.4/index.js
import "autotool-plugin";
import { join } from "node:path";
const name = "@alexaegis/autotool-plugin-vscode";
const description = "My vscode config";
const version = "0.8.4";
const license = "MIT";
const archetype = {
  platform: "node",
  framework: "autotool",
  language: "ts",
  kind: "lib"
};
const keywords = [
  "autotool-plugin",
  "managed-by-autotool",
  "vscode"
];
const type = "module";
const publishConfig = {
  access: "public"
};
const scripts = {
  build: "turbo run build-lib_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-vscode",
  "build-lib_": "vite build",
  "lint:depcheck": "turbo run lint:depcheck_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-vscode",
  "lint:depcheck_": "depcheck",
  "lint:es": "turbo run lint:es_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-vscode",
  "lint:es_": "eslint --max-warnings=0 --fix --no-error-on-unmatched-pattern .",
  "lint:format": "turbo run lint:format_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-vscode",
  "lint:format_": "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --check .",
  "lint:md": "turbo run lint:md_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-vscode",
  "lint:md_": "remark --frail --no-stdout --silently-ignore .",
  "lint:tsc": "turbo run lint:tsc_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-vscode",
  "lint:tsc_": "tsc --noEmit",
  publint: "BUILD_REASON='publish' turbo run publint_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-vscode",
  publint_: "publint dist",
  all: "BUILD_REASON='publish' turbo run all_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-vscode",
  format: "turbo run format_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-vscode",
  format_: "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --write .",
  test: "turbo run test_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-vscode",
  test_: "vitest --passWithNoTests --coverage --run",
  "test:watch": "vitest --passWithNoTests --coverage"
};
const exports = {
  ".": {
    types: "./src/index.ts",
    "import": "./dist/index.js",
    require: "./dist/index.cjs",
    "default": "./dist/index.js"
  },
  "./extensions": "./static/extensions.json",
  "./launch": "./static/launch.json",
  "./readme": "./readme.md",
  "./settings": "./static/settings.json",
  "./tasks": "./static/tasks.json"
};
const dependencies = {
  "@alexaegis/logging": "^0.7.2",
  "autotool-plugin": "^0.3.0"
};
const devDependencies = {
  "@alexaegis/eslint-config-vitest": "^0.8.4",
  "@alexaegis/ts": "^0.8.4",
  "@alexaegis/vite": "^0.8.4",
  "@alexaegis/vitest": "^0.8.4",
  "@types/node": "^20.9.0",
  publint: "^0.2.5",
  typescript: "^5.2.2",
  vite: "^4.5.0",
  vitest: "^0.34.6"
};
const packageJson = {
  name,
  description,
  version,
  license,
  "private": false,
  archetype,
  keywords,
  type,
  publishConfig,
  scripts,
  exports,
  dependencies,
  devDependencies
};
const plugin = (_options) => {
  return {
    name: packageJson.name,
    elements: [
      {
        description: "vscode settings",
        executor: "fileCopy",
        packageKind: "root",
        sourcePluginPackageName: packageJson.name,
        formatWithPrettier: true,
        sourceFile: join("static", "settings.json"),
        targetFile: join(".vscode", "settings.json")
      },
      {
        description: "vscode extension recommendations",
        executor: "fileCopy",
        packageKind: "root",
        sourcePluginPackageName: packageJson.name,
        formatWithPrettier: true,
        sourceFile: join("static", "extensions.json"),
        targetFile: join(".vscode", "extensions.json")
      },
      {
        description: "vscode launch configurations",
        executor: "fileCopy",
        packageKind: "root",
        sourcePluginPackageName: packageJson.name,
        formatWithPrettier: true,
        sourceFile: join("static", "launch.json"),
        targetFile: join(".vscode", "launch.json")
      },
      {
        description: "vscode tasks",
        executor: "fileCopy",
        packageKind: "root",
        sourcePluginPackageName: packageJson.name,
        formatWithPrettier: true,
        sourceFile: join("static", "tasks.json"),
        targetFile: join(".vscode", "tasks.json")
      }
    ]
  };
};
export {
  plugin as default,
  plugin
};
