// @alexaegis/autotool-plugin-turbo@0.8.4/index.cjs
"use strict";
Object.defineProperties(exports, { __esModule: { value: true }, [Symbol.toStringTag]: { value: "Module" } });
require("autotool-plugin");
const node_path = require("node:path");
const name = "@alexaegis/autotool-plugin-turbo";
const description = "My turbo config";
const version = "0.8.4";
const license = "MIT";
const archetype = {
  platform: "node",
  framework: "autotool",
  language: "ts",
  kind: "lib"
};
const keywords = [
  "autotool-plugin",
  "managed-by-autotool",
  "turbo"
];
const type = "module";
const publishConfig = {
  access: "public"
};
const scripts = {
  build: "turbo run build-lib_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-turbo",
  "build-lib_": "vite build",
  "lint:depcheck": "turbo run lint:depcheck_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-turbo",
  "lint:depcheck_": "depcheck",
  "lint:es": "turbo run lint:es_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-turbo",
  "lint:es_": "eslint --max-warnings=0 --fix --no-error-on-unmatched-pattern .",
  "lint:format": "turbo run lint:format_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-turbo",
  "lint:format_": "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --check .",
  "lint:md": "turbo run lint:md_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-turbo",
  "lint:md_": "remark --frail --no-stdout --silently-ignore .",
  "lint:tsc": "turbo run lint:tsc_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-turbo",
  "lint:tsc_": "tsc --noEmit",
  publint: "BUILD_REASON='publish' turbo run publint_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-turbo",
  publint_: "publint dist",
  all: "BUILD_REASON='publish' turbo run all_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-turbo",
  format: "turbo run format_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-turbo",
  format_: "prettier --cache-location .cache/prettier --plugin prettier-plugin-svelte --plugin prettier-plugin-tailwindcss --write .",
  test: "turbo run test_ --concurrency 16 --cache-dir .cache/turbo --filter @alexaegis/autotool-plugin-turbo",
  test_: "vitest --passWithNoTests --coverage --run",
  "test:watch": "vitest --passWithNoTests --coverage"
};
const exports$1 = {
  ".": {
    types: "./src/index.ts",
    "import": "./dist/index.js",
    require: "./dist/index.cjs",
    "default": "./dist/index.js"
  },
  "./readme": "./readme.md",
  "./turbo": "./static/turbo.json"
};
const dependencies = {
  "@alexaegis/logging": "^0.7.2",
  "autotool-plugin": "^0.3.0"
};
const devDependencies = {
  "@alexaegis/eslint-config-vitest": "^0.8.4",
  "@alexaegis/ts": "^0.8.4",
  "@alexaegis/vite": "^0.8.4",
  "@alexaegis/vitest": "^0.8.4",
  "@types/node": "^20.9.0",
  publint: "^0.2.5",
  turbo: "^1.10.16",
  "turbo-ignore": "^1.10.16",
  typescript: "^5.2.2",
  vite: "^4.5.0",
  vitest: "^0.34.6"
};
const packageJson = {
  name,
  description,
  version,
  license,
  "private": false,
  archetype,
  keywords,
  type,
  publishConfig,
  scripts,
  exports: exports$1,
  dependencies,
  devDependencies
};
const plugin = (_options) => {
  return {
    name: packageJson.name,
    elements: [
      {
        description: "workspace turbo config file",
        executor: "fileCopy",
        packageKind: "root",
        sourceFile: node_path.join("static", "turbo.json"),
        formatWithPrettier: true,
        targetFile: "turbo.json",
        sourcePluginPackageName: packageJson.name
      },
      {
        description: "package turbo scripts and devDependencies",
        executor: "packageJson",
        packageKind: "regular",
        data: {
          scripts: {
            // When checking everything, publint will only succeed when building for publishing
            all: "BUILD_REASON='publish' turbo run all_ --concurrency 16 --cache-dir .cache/turbo --filter ${packageName}"
          }
        }
      },
      {
        description: "workspace turbo scripts and devDependencies",
        executor: "packageJson",
        packageKind: "root",
        data: {
          scripts: {
            build: "turbo run build-lib_ build-app_ --concurrency 16 --cache-dir .cache/turbo",
            "build-lib": "turbo run build-lib_ --concurrency 16 --cache-dir .cache/turbo",
            all: "BUILD_REASON='publish' turbo run all_ --concurrency 16 --cache-dir .cache/turbo",
            lint: "turbo run lint_ --concurrency 16 --cache-dir .cache/turbo"
          },
          devDependencies: {
            turbo: packageJson.devDependencies.turbo,
            "turbo-ignore": packageJson.devDependencies["turbo-ignore"]
          }
        }
      }
    ]
  };
};
exports.default = plugin;
exports.plugin = plugin;
